<b>Creating clean repeatable code is a key part of a programmer. <br><br>
Functions allows us to create a block of code that can be executed many times, without needing to rewrite the code many times <br><br>
Syntax:
<br>
def function_name([arg1,...,argn]): 
    <br>'''Docstring explains code''' 
    <br>[return value] 
<br>variable_name = function_name([val1,...,valn])
<br><br>
Note: Anything enclosed in [ ] (square bracket) is optional.
</b>

In [1]:
def hello(): #function name should be all lower case
    '''
    This is the hello world code
    '''
    print("Hello Mohammad!")


In [2]:
hello()

Hello Mohammad!


In [5]:
def hello(name): #function name should be all lower case
    '''
    This is the hello world code
    '''
    print("Hello",name)


In [6]:
hello("Everyone!")

Hello Everyone!


<b> We use return keyword to send back the result of the function, instead of printing it out.<br>
    return allows us to assign the output to a new variable however print not allow us to do so.
</b>

In [7]:
def addition(num1,num2):
    return num1+num2

In [8]:
addition(2,3)

5

In [9]:
def hello(name='Mohammad!'): #passing default argument to avoid error
    '''
    This is the hello world code
    '''
    print("Hello",name)

In [10]:
hello()

Hello Mohammad!


In [11]:
hello("Rashid!")

Hello Rashid!


<b>Find out if the word 'dog' in the string</b>

In [12]:
def dog_check(mystr):
    if 'dog' in mystr.lower():
        return True
    else:
        return False

In [13]:
dog_check('He is a DOG Lover')

True

In [14]:
dog_check('He is a cat Lover')

False

<b>Password is valid or not
</b>

In [29]:
def passport_check(passport_no):
    if(len(passport_no)==8):
        if(passport_no[0]>="A" and passport_no[0]<="Z"):
            status="valid"
        else:
            status="invalid"
    else:
        status= "invalid"
    return(f"Passport is {status}")


In [30]:
passport_check("01234567")

'Passport is invalid'

In [31]:
passport_check("A1234567")

'Passport is valid'

<b>Write a Python function, find_square() that accepts an integer number n and returns the square of n. Invoke the function and display the square of the number. </b>

In [35]:
def find_square(n):
    return pow(n,2)

In [36]:
find_square(10)

100

In [37]:
def find_square(n):
    return pow(n,2)
num = int(input("Enter an Integer: "))
print("Square = ", find_square(num))

Enter an Integer: 10
Square =  100


<b>Write a Python function, find_sum() that accepts an integer n and returns the sum of first n numbers. Invoke the function and display the sum of first n numbers. </b>

In [45]:
def find_sum(n):
    summ=0
    i=1
    while i<=n:
        summ+=i
        i+=1
    return summ
num = int(input("Enter a number: "))
print(f"Sum of 1 to {num} is:",find_sum(num))

Enter a number: 10
Sum of 1 to 10 is: 55


In [46]:
find_sum(10
    )

55

<b>Driving age</b>

In [48]:
def checkDriverAge():
    age = input("What is your age?: ")
    if int(age) < 18:
        print("Sorry, you are too young to drive this car. Powering off")
    elif int(age) > 18:
        print("Powering On. Enjoy the ride!");
    elif int(age) == 18:
        print("Congratulations on your first year of driving. Enjoy the ride!")
checkDriverAge()

What is your age?: 188
Powering On. Enjoy the ride!


In [49]:
def checkDriverAge(age=0):
    if int(age) < 18:
        print("Sorry, you are too yound to drive this car. Powering off")
    elif int(age) > 18:
        print("Powering On. Enjoy the ride!");
    elif int(age) == 18:
        print("Congratulations on your first year of driving. Enjoy the ride!")


In [50]:
checkDriverAge(
    0)

Sorry, you are too yound to drive this car. Powering off


In [51]:
checkDriverAge(20)

Powering On. Enjoy the ride!


<b>Print Highest Even Number from given list</b>

In [104]:
def highest_even(li):
    evens = []
    for item in li:
        if item%2==0:
            evens.append(item)
    return max(evens)," is a maximum number in the list"
    
        

In [105]:
highest_even([10,8,2,3])

(10, ' is a maximum number in the list')

<b>
    '*args <br><br>
     “We use the “wildcard” or “*” notation like this – *args OR **kwargs – as our function’s argument when we have doubts about the number of  arguments we should pass in a function.” <br>we want to make a multiply function that takes any number of arguments and able to multiply them all together. It can be done using *args. It returns Tuple.<br><br>
    **kwargs<br><br>
    **kwargs in function definitions in python is used to pass a keyworded, variable-length argument list. We use the name kwargs with the double star. The reason is because the double star allows us to pass through keyword arguments (and any number of them).<br>It returns Dictionary.
</b>

In [63]:
#returns sum of given numbers
def myfunc(a,b):
    return a+b

In [64]:
myfunc(40,60)

100

In [65]:
myfunc(40,60,20)

TypeError: myfunc() takes 2 positional arguments but 3 were given

In [68]:
def myfunc2(*args):
    print(args) #returns tuple
    return sum(args)

In [69]:
myfunc2(40,60,20)

(40, 60, 20)


120

In [85]:
def myfunc3(*args,**kwargs):
    print(args) #returns tuple
    print(kwargs) #returns dictionary
    total=0
    for values in kwargs.values():
        total+=values
    return sum(args)+total

In [86]:
myfunc3(40,60,20,num1=2,num2=4)

(40, 60, 20)
{'num1': 2, 'num2': 4}


126

<b>walrus operator :=<br><br>
    The := operator is officially known as the assignment expression operator. <br>
    
</b>



In [108]:
a="hellllloooooooo"
if (length = len(a))>10:
    print(f"{len(a)} too long!")

SyntaxError: invalid syntax (Temp/ipykernel_8796/3219700664.py, line 2)

In [110]:
a="hellllloooooooo"
if (length := len(a))>10:
    print(f"{length} too long!")

15 too long!


<b>Scope - What variables do i have access to?
    <br> Rules:<br> 1. Starts with local scope 2. if there is no local then Parent local scope 3. if there is no Parent local then global scope 4. Built-in Functions eg. sum
    
</b>

In [118]:
a = 1 #3.global scope

def parent():
    a=10 #2.parent scope
    def confusion():
        a=5 #1.local scope
        return a
    return confusion()

print(a)
#print confusion()
print(parent()) #if we remove local scope a=5 it will return parent local a=10
print(a)

1
10
1


<b>Global keyword</b>

In [122]:
total = 0
def count():
    global total #we can do this better by doing this see below 
    total += 1
    return total

In [123]:
count()
count()
count()

3

In [127]:
total = 0
def count(total):
    
    total += 1
    return total

In [129]:
count(count(count(total)))


3

<b>Nonlocal keyword</b>

In [130]:
def outer():
    x = "local"
    def inner():
        nonlocal x
        x = "nonlocal"
        print("inner:", x)
    inner()
    print("outer:", x)
outer()


inner: nonlocal
outer: nonlocal


In [131]:
def outer():
    x = "local"
    def inner():
        #nonlocal x
        x = "nonlocal"
        print("inner:", x)
    inner()
    print("outer:", x)
outer()

inner: nonlocal
outer: local


In [132]:
wt_limit=30
def baggage_check(baggage_wt):
    extra_baggage_charge=0
    if not(baggage_wt>=0 and baggage_wt<=wt_limit):
        extra_baggage=baggage_wt-wt_limit
        extra_baggage_charge=extra_baggage*100
    return extra_baggage_charge
def update_baggage_limit(new_wt_limit):
    wt_limit=new_wt_limit
    print("This airline now allows baggage limit till",wt_limit,"kgs")
print("This airline allows baggage limit till",wt_limit,"kgs")
print("Pay the extra baggage charge of",baggage_check(35),"rupees")
update_baggage_limit(45)
print("Pay the extra baggage charge of",baggage_check(35),"rupees")


This airline allows baggage limit till 30 kgs
Pay the extra baggage charge of 500 rupees
This airline now allows baggage limit till 45 kgs
Pay the extra baggage charge of 500 rupees


<b>Write a Python function factorial(num) which returns the factorial of a given number.</b>

In [1]:
def factorial(num):
    fact = 1
    if num < 0:
       print("Sorry, factorial does not exist for negative numbers")
    elif num == 0:
       print("The factorial of 0 is 1")
    else:
       for i in range(1,num + 1):
           fact = fact*i
       print("The factorial of",num,"is",fact)

In [2]:
factorial(5)

The factorial of 5 is 120


<b> Write a Python Function is_palindrome(num) that accepts an integer num as argument and returns True if the num is palindrome else returns false. Invoke the function and based on return value print the output.

Example: num=12321 output: Given number is a palindrome, num=12345 output: Given number is not a palindrome </b>

In [11]:
def is_palindrome(num):
    temp = num
    Reverse = 0
    while(num > 0):
        dig = num%10
        Reverse = Reverse *10 + dig
        num = num//10
    if temp == Reverse:
        return True
    else:
        return False
    
n = int(input("Enter Number: "))
print(f"{n} is {is_palindrome(n)}")


Enter Number: 12332
12332 is False


<b> Write a Python function check_amicable_numbers(num1, num2) that accepts two numbers num1 and num2 as arguments and returns True if the given pair of numbers are amicable numbers else return false. Invoke the function and based on return value print the numbers are amicable numbers or not.

num1 and num2 are said to be amicable numbers if sum of all the proper devisors (except num1 itself) of num1 is equal to num2 and sum of all the proper devisors of num2 (except num1 itself) is equal to num1.

Example: 220 and 284 are amicable numbers as

Proper devisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55, 110 whose sum is 284

Proper devisors of 284 are 1, 2, 4, 71, 142 whose sum is 220 </b>

In [14]:
def check_amicable_numbers(num1, num2):
    sum1 = 0
    sum2 = 0
    for i in range(1,num1):
        if num1%i==0:
            sum1+=i
    for j in range(1,num2):
        if num2%j==0:
            sum2+=j
    if(sum1==num2 and sum2==num1):
        return f"{num1} and {num2} are amicable numbers"
    else:
        return f"{num1} and {num2} are not amicable numbers"

n1 = int(input("Enter Number 1: "))
n2 = int(input("Enter Number 2: "))
print(check_amicable_numbers(n1, n2))

Enter Number 1: 220
Enter Number 2: 284
220 and 284 are amicable numbers


<b> Write a Python function right_shift(num, n) that takes two numbers num and n as  arguments and returns value of the integer num rotated to the right by n positions. Assume both the numbers are unsigned. Invoke the function and print the return value.

Hint: use >> binary operator to shift the bits.

Example: num=60, n=2 result=15 </b>

In [16]:
def right_shift(num, n):
    if num>=0 and n>=0:
        return num >> n
    else:
        return "Signed Number"
    
a = int(input("Enter Number 1: "))
n = int(input("Enter Number 2: "))
print(f"result: {right_shift(a, n)}")
    

Enter Number 1: 60
Enter Number 2: 2
result: 15


<b> Write a Python function check_strong_number(num) that accepts a positive integer as argument and returns True if the number is strong number else false. Invoke the function and based on return value print the number is strong number or not.

A number is said to be strong number, if the sum of factorial of each digit of the number is equal to the given number.

Example:145 is strong number as

1!=1

4!=24

5!=120

Sum of all these is 145 </b>

In [20]:
def check_strong_number(num):
    sum1=0
    temp=num
    while(num):
        i=1
        factorial=1
        r=num%10
        while(i<=r):
            factorial=factorial*i
            i=i+1
        sum1=sum1+factorial
        num=num//10
    if(sum1==temp):
        return "The number is a strong number"
    else:
        return "The number is not a strong number"
        
n = int(input("Enter Number: "))
print(check_strong_number(n)) 
    

Enter Number: 145
The number is a strong number


<b>Observe & Predict the output of the following code snippet.</b>

In [5]:
def verify(num1,num2):
    if num1 > num2:
        return num1
    elif num1 == num2:
        return 1
    else:
        return num2

def display(arg1,arg2):
    if(verify(arg1,arg2)==arg1):
        print("A")
    elif(verify(arg1,arg2)==1):
        print("C")
    else:
        print("B")
display(1000,3500)

B


In [6]:
result=0

def find_sum(num1,num2):
    if(num1!=num2):
        result=num1+num2
    else:
        result=2*(num1+num2)

find_sum(3,4)
print(result)
find_sum(5,5)
print(result)

0
0


In [15]:
def find_avg(list_num):
    result_sum=0
    for num in list_num:
        result_sum+=num
    result_avg=result_sum/len(list_num)
find_avg([5,8,5])
print(result_avg)
 

NameError: name 'result_avg' is not defined

<b>Write a Python function proper_divisors(num) which returns a list of all the proper divisors of given number.

Example: Proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55, 110</b>

In [5]:
def proper_divisors(num):
    mylist=[]
    i=1
    while i<num:
        if num%i == 0:
            mylist.append(i)
        
        i+=1
    return mylist

In [6]:
proper_divisors(220)

[1, 2, 4, 5, 10, 11, 20, 22, 44, 55, 110]

<b>Write a Python function generate_fibanacci(n) to return a list of first n Fibonacci numbers.</b>

In [21]:
def fib(n):
    if n == 0:
        return [0]
    elif n == 1:
        return [0, 1]
    else:
        lst = fib(n-1)
        lst.append(lst[-1] + lst[-2])
        return lst

In [22]:
fib(8)

[0, 1, 1, 2, 3, 5, 8, 13, 21]

<b>
Given a string containing uppercase characters (A-Z), compress the string using Run Length encoding. Repetition of character has to be replaced by storing the length of that run.

Write a python function encode(message)  which performs the run length encoding for a given String and returns the run length encoded String.

Provide different String values and test your program.

Example: message=AAAABBBBCCCCCCCC  output: 4A4B8C</b>

In [64]:
def encode(message):
    pairs = []
    for char in message:
        if len(pairs) > 0:
            if pairs[-1][0] == char:
                pairs[-1] = (char, pairs[-1][1] + 1)
            else:
                pairs.append((char, 1))
        else:
            pairs.append((char, 1))
    strings = []
    for letter, count in pairs:
        strings.append(f"{count}{letter.upper()}")
    return "".join(strings)


In [65]:
encode("AAAABBBBCCCCCCCC")

'4A4B8C'

<b>
    Represent a small bilingual (English-Swedish) glossary given below as a Python dictionary

{"merry":"god", "christmas":"jul", "and":"och", "happy":"gott", "new":"nytt", "year":"ar"} 

and use it to translate your Christmas wishes from English into Swedish.

That is, write a Python function translate(b_dict,list_words) that accepts the bilingual dictionary and a list of English words (your Christmas wish) and returns a list of equivalent Swedish words. </b>

In [70]:
def translate(b_dict,list_words):
    swedish_words_list=[]
    for x in list_words:
        swedish_words_list.append(b_dict[x])
    return swedish_words_list
    
b_dict= {"merry":"god", "christmas":"jul", "and":"och", "happy":"gott", "new":"nytt", "year":"ar"}
list_words = ["merry","christmas"]
swedish_words_list=translate(b_dict,list_words)
print("The English Words are:",list_words)
print("The equivalent swedish words are:",swedish_words_list)

The English Words are: ['merry', 'christmas']
The equivalent swedish words are: ['god', 'jul']
